### PR TITLE
[BEAM-5866] Fix `Row#equals`

### PR DESCRIPTION
**Pull request by kanterov**
This pull request defines two properties that any `Coder` implementation should satisfy. There are two coders that don't properly implement `structuralValue`: `MapCoder` and `RowCoder`.

`RowCoder#consistentWithEquals` is defined as always true. It isn't true in case schema has field of type `Map<BYTES, ?>`. However, such field type makes little sense. One of the ways to fix it would be to reject such schemas. For now, this test is ignored.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by kanterov** [comment]
R: @reuvenlax @amaliujia 

Can you please take a look, it's a problem similar to https://github.com/apache/beam/pull/6765

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderTestUtils.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+package org.apache.beam.sdk.coders;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assume.assumeTrue;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+
+public class CoderTestUtils {
```
Comment (kanterov) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/CoderTestUtils.java]: Just discovered there is `CoderProperties` with such functions, probably it should go there

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/org/apache/beam/sdk/coders/RowCoderTest.java [hunk 2]
```diff
@@ -145,4 +149,47 @@ public void testVerifyDeterministicNestedRow() throws NonDeterministicException
...
+  public void testStructuralValueBytesField() {
+    Schema schema = Schema.of(Schema.Field.of("f1", FieldType.BYTES));
+    Row row = Row.withSchema(schema).addValue(new byte[] {1, 2, 3, 4}).build();
+    RowCoder coder = RowCoder.of(schema);
+
+    CoderTestUtils.testStructuralValueConsistentWithEquals(coder, row);
+  }
+
+  @Test
+  @Ignore // TODO should it even be possible to create rows with such schema?
```
Comment (kanterov) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/coders/org/apache/beam/sdk/coders/RowCoderTest.java]: Need advice what to do

**Comment by kanterov** [comment]
One of the ways to fix would be to reduce the number of allocations in `structuralValue` (already done) and use it in `Row#equals` instead.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/org/apache/beam/sdk/coders/RowCoderTest.java (hunk 2 above)
Comment (kennknowles) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/coders/org/apache/beam/sdk/coders/RowCoderTest.java]: This seems like an OK schema to me. What is the problem with it?

File: sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoder.java [hunk 3]
```diff
@@ -136,6 +137,20 @@ private void verifyDeterministic(Schema schema)
     Coder.verifyDeterministic(this, "All fields must have deterministic encoding", coders);
   }
 
+  @Override
+  public Object structuralValue(final Row row) {
```
Comment (kennknowles) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoder.java]: I think making `Row.equals` a custom deep equality (so byte arrays with the same contents are equal, etc) is the best solution here. Then it is always consistent with equals and no structuralValue is needed. Is there a pitfall?

Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoder.java]: That doesn't really make sense. The comparisons you have to do for correctness are the same, and what I'm suggestion would eliminate the copying/allocation.

Comment (kennknowles) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/coders/RowCoder.java]: Row still has incorrect hashcode and equals so DirectRunner would actually just shuffle wrong if there were any BYTES field. Which runner was it?

File: sdks/java/core/src/test/java/org/apache/beam/sdk/coders/org/apache/beam/sdk/coders/RowCoderTest.java (hunk 2 above)
Comment (kanterov) [reply sdks/java/core/src/test/java/org/apache/beam/sdk/coders/org/apache/beam/sdk/coders/RowCoderTest.java]: `Map` implementations imply that there is either order or hashCode+equality for values. `byte[]`doesn't have either of that. `Map<byte[], ?>` will work only if we provide specialized implementation for it.

**Comment by kanterov** [comment]
@reuvenlax @kennknowles Thanks for the review. I tried to fix `Row#equals` with custom `deepEquals`, but it turns out to be terrible in cases of `List<byte[]>`, `List<List<byte[]>>`, `Map<?, byte[]>`, `Map<byte[], ?>`, etc., because it requires traversing row schema, or using a lot of reflection due to type erasure.

I did a different thing, stored `byte[]` as `StructuralByteArray`, and `deepEquals` isn't needed anymore. One drawback is that `Row#getValue` changes return type of `BYTES` field.

I have a few suggestions I want to discuss:
- use `ByteBuffer` instead of `StructuralByteArray` to reduce the amount of "custom" types
- store `BYTES` as `byte[]` instead of `StructuralByteArray` in the POJO-backed implementation of `Row` to reduce extra overhead

**Comment by kennknowles** [comment]
Whatever implementation we use, it has to have the same semantics as the custom deep equals. Two rows are equal if they are observably the same when limited to the properties of the type in their schema. The representation should be able to change. We can ignore Java's broken array equality except as an example of something to not do. And since this equality is already going to be consistent with serialization-based shuffling, we shouldn't need structural value (I'm OK with it, but it is surprising to need it). I am surprised, and sort of not believing, that a schema-driven deep equality could be slower than a recursive structuralValue. And I also think that it is possible to have correct semantics with good performance.

**Comment by kanterov** [comment]
@reuvenlax Yes, the problem with ByteBuffer that it has too many fields we don't need, like `positions`, `limit`, `offset`, `mark`, `capacity`, `isReadonly`. They make sense when implementing protocols, but aren't so much useful in data pipelines.

@kennknowles do we even want hashmap as implementation? In Spark, it's done with ArrayBasedMaps. I was thinking similar could be done, for instance, for arrays, to use primitive non-boxing collections where applicable.

**Comment by kennknowles** [comment]
SGTM to try different map implementations. I would like to preserve our ability to change the decision of what map data structure is used in a Java implementation of rows without any breaking change for users. It may just need to be altered over time, or it may need to be tunable based on application.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java [hunk 4]
```diff
@@ -346,13 +346,126 @@ public boolean equals(Object o) {
...
+      }
+
+      if (a.size() != b.size()) {
+        return false;
+      }
+
+      for (Map.Entry<K, V> e : a.entrySet()) {
+        K key = e.getKey();
+        V value = e.getValue();
+        V otherValue = b.get(key);
```
[the remark this answers is not in the document]
Comment (amaliujia) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java]: And given this idea, if change the `deepHashCode` to the version that compute hashcode of Map's byte[] key based on Object but not array content, `deepHashCode` and `deepEquals` will be consistent.

**Comment by kanterov** [comment]
@kennknowles I was thinking about this problem over the weekend. Everything becomes more simple with `OurBytesTypeWithGoodEquals`. I pushed the implementation in https://github.com/kanterov/beam/commit/08f300e800ef056238360dcea33a47f95ddb5d3d. I think overall it's the best approach.

**API** It preserves the ability to change internal representation without any surprises or breaking APIs.

**Correctness** It's clear from the beginning what is `Map<OurBytesTypeWithGoodEquals, T>`, no matter how it's implemented. `Map<byte[], ?>` is tricky, and trivial implementation is buggy.

**Performance** I did JMH benchmarks, it's similar to `deepEquals`:
```
RowBenchmark.rowWithStorageAndDeepEquals        thrpt   10  0.014 ± 0.001  ops/ns
RowBenchmark.rowWithStorageAndByteArrayWrapper  thrpt   10  0.015 ± 0.001  ops/ns
RowBenchmark.rowWithGetterAndPojoEquals         thrpt   10  0.022 ± 0.001  ops/ns
```

Interesting learning is that the most performant approach would be generating POJOs based on a schema, this could be done without breaking proposed API.

**Simplicity** `Row` implementation is more simple because it doesn't need custom `deepEquals`, or specialized collection implementations.

**Map<BYTES, ?>** It's naturally supported without handling it as a special case

**Comment by kennknowles** [comment]
That really sounds quite good to me. Even without schemas it makes sense, because shuffling based on serialized keys means that `byte[]` equality always needed a wrapper to be in-memory-compatible with shuffle. At this point, we have done far too much on a PR. Would you bring your latest comment to the dev list for a discussion?

**Comment by reuvenlax** [comment]
Changing Row.get to return StructuralByteArray instead of byte[] will
probably require some changes to the byte-buddy generated code.

**Comment by kanterov** [comment]
@reuvenlax yes, I did it in https://github.com/kanterov/beam/commit/08f300e800ef056238360dcea33a47f95ddb5d3d, quite amazing stuff in byte buddy btw :).

**Comment by reuvenlax** [comment]
Is this waiting for further review?

**Comment by kanterov** [comment]
@reuvenlax @amaliujia In my opinion, this PR is a step forward because it fixes few tests, but we don't have a decision if we want to proceed with deep equality vs. changing the representation of `BYTES` to `StructuralByteArray`.

We have arguments for and against using `StructuralByteArray`, and an example implementation, but I don't understand how decisions are made, so it isn't clear how to proceed.

However, the absence of decision doesn't block this PR, because it extends deep equals approach that we use today, therefore, in my opinion, we can merge it with the approval of Beam committers.

**Comment by kennknowles** [comment]
Sorry for the delay - this LGTM.

**Comment by kennknowles** [comment]
We can move forward and continue to consider the semantics of schemas on list.

**Pull request merged**